README.md
# NAME

Redis::LeaderBoard - leader board using Redis

# SYNOPSIS

    use Redis;
    use Redis::LeaderBoard;
    my $redis = Redis->new;
    my $lb = Redis::LeaderBoard->new(
        redis => $redis,
        key   => 'leader_board:1',
        order => 'asc', # asc/desc, desc as default
    );
    $lb->set_score('one' => 100');
    $lb->set_score('two' =>  50');
    my ($rank, $score) = $lb->get_rank_with_score('one');

    # memmber object
    my $member = $lb->find_member('two');
    $member->score(90);
    my $rank2 = $member->rank;

# DESCRIPTION

Redis::LeaderBoard is for providing leader board by using Redis's sorted set.

z(rev)?rank of Redis 2.8 or older doesn't consider same scores.
This module resolve it.

**THIS IS A ALPHA QUALITY RELEASE. API MAY CHANGE WITHOUT NOTICE**.

# INTERFACE

## Constructor

### `my $lb = Redis::LeaderBoard->new(%options)`

Create a new leader board object. Options should be set in `%options`.

- `redis: Redis`

    Redis object. Redis.pm or Redis::hiredis.

- `key: Str`

    Required.

- `order: Enum(asc/desc)`

    Optional. `desc` as default.

## Methods

### `$member_obj:Redis::LeaderBoard::Member = $lb->find_member($member:Str)`

Find member by member id. see [Redis::LeaderBoard::Member](https://metacpan.org/pod/Redis::LeaderBoard::Member) for more details.

### `$lb->set_score($member:Str, $score:Number, [$member2, $score2,...])`

Set scores of members. You can set multiple element if using Redis 2.4 or later.

### `$score:Number = $lb->get_score($member:Str)`

Get score of member.

### `$score:Number = $lb->incr_score($member:Str, [$increment_score:Number])`

increment score of member and returns reflected score. 1 is default `$increment_score`.

### `$score:Number = $lb->decr_score($member:Str, [$decrement_score:Number])`

decrement score of member and returns reflected score. 1 is default `$decrement_score`.

### `$lb->remove($member:Str, [$member2:Str,...])`

remove members from leader board. Multiple element can be accepted Redis 2.4 or later.

### `($rank:Int, $score:Number) = $lb->get_rank_with_score($member:Str)`

Returns rank and score. If you want to get rank and score at the same time,
you should not call `get_score` and `get_rank` separately, use this method instead for
performance.

### `$rank:Int = $lb->get_rank($member:Str)`

Get rank of member.

### `$order:Int = $lb->get_sorted_order($member:Str)`

Get sorted order in sorted set. (same as `$redis->zrank`)

### `$count = $lb->member_count([$from, $to])`

Get number of members. If score range (`$from` and `$to`) is specified, it returns a number
of members in range.

### `$rankings:ArrayRef<HashRef> = $lb->rankings(%opt)`

Return rankings by arrayref contains hashrefs.
keys of hashref is `member:Str`, `rank:Int` and `score:Number`.

Options can be set in `%options`. keys of options are as follows.

- `limit: Int`
- `offset: Int`

# LICENSE

Copyright (C) Masayuki Matsuki.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Masayuki Matsuki <y.songmu@gmail.com>


Changes
Revision history for Perl extension Redis-LeaderBoard

{{$NEXT}}

1.11 2016-02-06T13:55:11Z
    - Make limit atomic (shogo82148++)

1.10 2015-11-23T16:22:42Z
    - add `limit` accessor

1.01 2015-09-30T11:52:00Z
    - add `get_rank_by_score` method and refactory

1.00 2015-09-29T16:18:10Z
    - optimize `ranking` method using 'WITHSCORES'

0.03 2014-02-03T08:23:10Z
    - add `expire_at` constructor option.

0.02 2013-12-30T03:54:16Z
    - `member_count` method can accept $from and $to arguments

0.01 2013-06-02T07:18:21Z
    - original version


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Redis-LeaderBoard',
    module_name     => 'Redis::LeaderBoard',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,


);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

use File::Copy;

print "cp META.json MYMETA.json\n";
copy("META.json","MYMETA.json") or die "Copy failed(META.json): $!";

if (-f 'META.yml') {
    print "cp META.yml MYMETA.yml\n";
    copy("META.yml","MYMETA.yml") or die "Copy failed(META.yml): $!";
} else {
    print "There is no META.yml... You may install this module from the repository...\n";
}



lib/Redis/Ranking.pm
package Redis::Ranking;
use 5.008001;
our $VERSION = "0.01";
use Mouse;
use Mouse::Util::TypeConstraints;

has key => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

has redis => (
    is       => 'ro',
    isa      => 'Object',
    required => 1,
);

enum 'Redis::Ranking::Order' => qw/asc desc/;
has order => (
    is      => 'ro',
    isa     => 'Redis::Ranking::Order',
    default => 'asc',
);

no Mouse;

# consider bulk?
sub set_score {
    my ($self, $member, $score) = @_;
    $self->redis->zadd($self->key, $score, $member);
}

sub get_score {
    my ($self, $member) = @_;
    $self->redis->zscore($self->key, $member);
}

sub incr_score {
    my ($self, $member, $score) = @_;
    $score = defined $score ? $score : 1;

    $self->redis->zincrby($self->key, $score, $member);
}

sub decr_score {
    my ($self, $member, $score) = @_;
    $score = defined $score ? $score : 1;

    $self->redis->zincrby($self->key, -$score, $member);
}

# consider bulk?
sub remove {
    my ($self, $member) = @_;

    $self->redis->zrem($self->key, $member);
}

sub get_sorted_order {
    my ($self, $member) = @_;

    my $method = $self->order eq 'asc' ? 'zrevrank' : 'zrank';
    $self->redis->$method($self->key, $member);
}

sub get_rank_with_score {
    my ($self, $member) = @_;
    my $redis = $self->redis;

    my $score = $self->get_score($member);
    return unless defined $score; # should throw exception?

    my $method = $self->order eq 'asc' ? 'zrevrank' : 'zrank';
    my $rank  = $self->get_sorted_order($member);

    return (1, $score) if $rank == 0; # zero origin

    my ($min, $max) = $self->order eq 'asc' ? ("($score", 'inf') : ('-inf', "($score");
    my $above_count = $redis->zcount($self->key, $min, $max);
    $rank = $above_count + 1;

    ($rank, $score);
}

sub get_rank {
    my ($self, $member) = @_;

    my ($rank) = $self->get_rank_with_score($member);
    $rank;
}

sub get_rankings {
    my ($self, %args) = @_;
    my $limit  = exists $args{limit}  ? $args{limit}  : $self->member_count;
    my $offset = exists $args{offset} ? $args{offset} : 0;

    my $range_method = $self->order eq 'asc' ? 'zrevrange' : 'zrange';

    my $members = $self->redis->$range_method($self->key, $offset, $offset + $limit - 1);
    my @rankings;
    # needs pipelie?
    for my $member (@$members) {
        my ($rank, $score) = $self->get_rank_with_score($member);

        push @rankings, +{
            member => $member,
            score  => $score,
            rank   => $rank,
        };
    }
    \@rankings;
}

sub member_count {
    my $self = shift;
    $self->redis->zcard($self->key);
}

1;
__END__

=encoding utf-8

=head1 NAME

Redis::Ranking - It's new $module

=head1 SYNOPSIS

    use Redis::Ranking;

=head1 DESCRIPTION

Redis::Ranking is ...

=head1 LICENSE

Copyright (C) Masayuki Matsuki.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Masayuki Matsuki E<lt>y.songmu@gmail.comE<gt>

=cut



lib/Redis/LeaderBoard/Member.pm
package Redis::LeaderBoard::Member;
use Mouse;

has leader_board => (
    is       => 'ro',
    isa      => 'Redis::LeaderBoard',
    required => 1,
);

has member => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

no Mouse;

sub score {
    my ($self, $score) = @_;

    return $self->leader_board->get_score($self->member) unless $score;

    $self->leader_board->set_score($self->member, $score);
    $score;
}

sub incr {
    my ($self, $score) = @_;
    $score = defined $score ? $score : 1;

    $self->leader_board->incr_score($self->member, $score);
}

sub decr {
    my ($self, $score) = @_;
    $score = defined $score ? $score : 1;

    $self->leader_board->decr_score($self->member, $score);
}

sub rank_with_score {
    my $self = shift;
    $self->leader_board->get_rank_with_score($self->member);
}

sub rank {
    my $self = shift;
    $self->leader_board->get_rank($self->member);
}

sub sorted_order {
    my $self = shift;

    $self->leader_board->sorted_order($self->member);
}

1;


lib/Redis/LeaderBoard.pm
package Redis::LeaderBoard;
use 5.008001;
our $VERSION = "1.10";
use Mouse;
use Mouse::Util::TypeConstraints;
use Redis::LeaderBoard::Member;

has key => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

has redis => (
    is       => 'ro',
    isa      => 'Object',
    required => 1,
);

enum 'Redis::LeaderBoard::Order' => qw/asc desc/;
has order => (
    is      => 'ro',
    isa     => 'Redis::LeaderBoard::Order',
    default => 'desc',
);

has is_asc => (
    is   => 'ro',
    isa  => 'Bool',
    lazy => 1,
    default => sub { shift->order eq 'asc' },
);

has expire_at => (
    is   => 'ro',
    isa  => 'Int',
);

has limit => (
    is  => 'ro',
    isa => 'Int',
);

no Mouse;

sub find_member {
    my ($self, $member) = @_;

    Redis::LeaderBoard::Member->new(
        member       => $member,
        leader_board => $self,
    );
}

sub set_score {
    my ($self, @member_and_scores) = @_;
    @member_and_scores = reverse @member_and_scores;
    $self->redis->zadd($self->key, @member_and_scores);
    $self->_set_expire_and_limit;
}

sub get_score {
    my ($self, $member) = @_;
    $self->redis->zscore($self->key, $member);
}

sub incr_score {
    my ($self, $member, $score) = @_;
    $score = defined $score ? $score : 1;

    my $ret = $self->redis->zincrby($self->key, $score, $member);
    $self->_set_expire_and_limit;
    $ret;
}

sub decr_score {
    my ($self, $member, $score) = @_;
    $score = defined $score ? $score : 1;

    my $ret = $self->redis->zincrby($self->key, -$score, $member);
    $self->_set_expire_and_limit;
    $ret;
}

sub _set_expire_and_limit {
    my $self = shift;
    $self->redis->expireat($self->key, $self->expire_at) if $self->expire_at;

    if ($self->limit) {
        my ($from, $to) = (0, -$self->limit-1);
        if ($self->is_asc) {
            ($from, $to) = ($self->limit, -1)
        }
        $self->redis->zremrangebyrank($self->key, $from, $to);
    }
}

sub remove {
    my ($self, @members) = @_;

    $self->redis->zrem($self->key, @members);
}

sub get_sorted_order {
    my ($self, $member) = @_;

    my $method = $self->is_asc ? 'zrank' : 'zrevrank';
    $self->redis->$method($self->key, $member);
}

sub get_rank_with_score {
    my ($self, $member) = @_;
    my $redis = $self->redis;

    my $score = $self->get_score($member);
    return unless defined $score;

    my $rank = $self->get_rank_by_score($score);
    ($rank, $score);
}

sub get_rank {
    my ($self, $member) = @_;

    my ($rank) = $self->get_rank_with_score($member);
    $rank;
}

sub rankings {
    my ($self, %args) = @_;
    my $limit  = exists $args{limit}  ? $args{limit}  : $self->member_count;
    my $offset = exists $args{offset} ? $args{offset} : 0;

    my $range_method = $self->is_asc ? 'zrange' : 'zrevrange';

    my $members_with_scores = $self->redis->$range_method($self->key, $offset, $offset + $limit - 1, 'WITHSCORES');
    return [] unless @$members_with_scores;

    my @rankings;
    my ($current_rank, $current_target_score, $same_score_members);
    while (my ($member, $score) = splice @$members_with_scores, 0, 2) {
        if (!$current_rank) {
            $current_rank         = $self->get_rank_by_score($score);
            $same_score_members   = $offset - $current_rank + 2;
            $current_target_score = $score;
        }
        elsif ($score == $current_target_score) {
            $same_score_members++;
        }
        else {
            $current_target_score = $score;
            $current_rank = $current_rank + $same_score_members;
            $same_score_members = 1;
        }
        push @rankings, +{
            member => $member,
            score  => $score,
            rank   => $current_rank,
        };
    }

    \@rankings;
}

sub get_rank_by_score {
    my ($self, $score) = @_;

    my ($min, $max) = $self->is_asc ? ('-inf', "($score") : ("($score", 'inf');
    return $self->member_count($min, $max) + 1;
}

sub member_count {
    my ($self, $from, $to) = @_;

    if (!$from && !$to) {
        $self->redis->zcard($self->key);
    }
    else {
        $from = defined $from ? $from : '-inf';
        $to   = defined $to   ? $to   : 'inf';
        $self->redis->zcount($self->key, $from, $to);
    }
}

1;
__END__

=encoding utf-8

=head1 NAME

Redis::LeaderBoard - leader board using Redis

=head1 SYNOPSIS

    use Redis;
    use Redis::LeaderBoard;
    my $redis = Redis->new;
    my $lb = Redis::LeaderBoard->new(
        redis => $redis,
        key   => 'leader_board:1',
        order => 'asc', # asc/desc, desc as default
    );
    $lb->set_score('one' => 100');
    $lb->set_score('two' =>  50');
    my ($rank, $score) = $lb->get_rank_with_score('one');

    # memmber object
    my $member = $lb->find_member('two');
    $member->score(90);
    my $rank2 = $member->rank;

=head1 DESCRIPTION

Redis::LeaderBoard is for providing leader board by using Redis's sorted set.

z(rev)?rank of Redis 2.8 or older doesn't consider same scores.
This module resolve it.

=head1 INTERFACE

=head2 Constructor

=head3 C<< my $lb = Redis::LeaderBoard->new(%options) >>

Create a new leader board object. Options should be set in C<%options>.

=over

=item C<redis: Redis>

Redis object. Redis.pm or Redis::hiredis.

=item C<key: Str>

Required.

=item C<order: Enum(asc/desc)>

Optional. C<desc> as default.

=back

=head2 Methods

=head3 C<< $member_obj:Redis::LeaderBoard::Member = $lb->find_member($member:Str) >>

Find member by member id. see L<Redis::LeaderBoard::Member> for more details.

=head3 C<< $lb->set_score($member:Str, $score:Number, [$member2, $score2,...]) >>

Set scores of members. You can set multiple element if using Redis 2.4 or later.

=head3 C<< $score:Number = $lb->get_score($member:Str) >>

Get score of member.

=head3 C<< $score:Number = $lb->incr_score($member:Str, [$increment_score:Number]) >>

increment score of member and returns reflected score. 1 is default C<$increment_score>.

=head3 C<< $score:Number = $lb->decr_score($member:Str, [$decrement_score:Number]) >>

decrement score of member and returns reflected score. 1 is default C<$decrement_score>.

=head3 C<< $lb->remove($member:Str, [$member2:Str,...]) >>

remove members from leader board. Multiple element can be accepted Redis 2.4 or later.

=head3 C<< ($rank:Int, $score:Number) = $lb->get_rank_with_score($member:Str) >>

Returns rank and score. If you want to get rank and score at the same time,
you should not call C<get_score> and C<get_rank> separately, use this method instead for
performance.

=head3 C<< $rank:Int = $lb->get_rank($member:Str) >>

Get rank of member.

=head3 C<< $order:Int = $lb->get_sorted_order($member:Str) >>

Get sorted order in sorted set. (same as C<< $redis->zrank >>)

=head3 C<< $count = $lb->member_count([$from, $to]) >>

Get number of members. If score range (C<$from> and C<$to>) is specified, it returns a number
of members in range.

=head3 C<< $rankings:ArrayRef<HashRef> = $lb->rankings(%opt) >>

Return rankings by arrayref contains hashrefs.
keys of hashref is C<member:Str>, C<rank:Int> and C<score:Number>.

Options can be set in C<%options>. keys of options are as follows.

=over

=item C<limit: Int>

=item C<offset: Int>

=back

=head1 LICENSE

Copyright (C) Masayuki Matsuki.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Masayuki Matsuki E<lt>y.songmu@gmail.comE<gt>

=cut

